designs/pluggable-message-routing/README.md

```diff
@@ -0,0 +1,263 @@
+
+
+#### Life-Cycle Hooks
+
+This is similar to the Fastify model of registering hooks at certain key points
+in the life-cycle of a message.
+
+Rather than leave the stack as being completely customisable, we recognise there
+is a core set of steps that every message has to go through. Within those steps
+are points where some custom code *might* want to run.
+
+The following diagram shows a possible set of hook points. This list may well get
+refined as the design progresses. The orange lines show the span of synchronous calls.
+
+![](message-router-events.png)
+
+1. `preSend` - passed an array of `SendEvent` objects. The messages inside these objects
+   are exactly what the node has passed to `node.send` - meaning there could be duplicate
+   references to the same message object.
+
+2. `preRoute` - called once for each SendEvent object in turn
+
+3. `onSend` - the local router has identified the node it is going to send to. At
+   this point, the message has been cloned if needed.
+
+4. `postSend` - the message has been dispatched to be delivered asynchronously
+(unless the sync delivery flag is set, in which case it would be continue as synchronous delivery)
+
+5. `onReceive` - a node is about to receive a message
+
+6. `postReceive` - the message has been passed to the node's `input` handler
+
+7. `onDone`, `onError` - the node has completed with a message or logged an error
+
+
+*The exact function signature for each of these hooks needs to be designed.*
+
+They operate on `SendEvent` objects. Each objects contains the message and metadata
+needed through the process.
+
+They start out looking like:
+
+```json
+{
+    "msg": "<message object>",
+    "source": {
+        "id": "<node-id>",
+        "node": "<node-object>",
+        "port": "<index of port being sent on>",
+    },
+    "destination": {
+        "id": "<node-id>",
+        "node": undefined,
+    },
+    "cloneMessage": "true|false"
+}
+```
+
+##### Scenario: custom cloning behaviour
+
+The initial logic around what messages get cloned does not change - the node
+identifies if cloning needs to occur based on how many nodes it is wired to and
+how many messages it is asked to send in a single go.
+
+Rather than do the cloning, the node will set the `cloneMessage` property to indicated
+whether cloning should occur or not.
+
+The actual cloning would happen between the `preRoute` and `onSend` steps.
+
+This would allow a `preRoute` handler to do its own cloning behaviour and then set
+`cloneMessage` to `false` so that no further cloning would happen for that message.
+
+One very important aspect to highlight is that messages will not be cloned until
+after the `preRoute` stage - so any modifications of the message will potentially
+mutate multiple messages. If earlier handlers want to modify the message in any
+way, they will have to honour the `cloneMessage` flag and clone the message themselves.
+
+
+##### Scenario: remote message routing
+
```

Comment:
in this instance - does destination just have more properties such that it knows what to do - or is the `node` property replaced by a "remote" target definition ?

Reply:
At the `preRoute` step, we only have the `id` of the node it has to send to. It is up to the handler to decide what it means to send to that `id`. Other runtime APIs would need to be provided to let it 'see' the full flow configuration and determine if any given `id` is local or remote and if there is any additional meta data embedded in the flow (such as group properties). Those other APIs still need to be designed.

Reply:
Ok